tests/test_optimizer.py

```diff
@@ -88,6 +88,27 @@ def qualify_columns(expression, **kwargs):
 
         self.check_file("qualify_columns", qualify_columns, schema=self.schema)
 
+    def test_qualify_columns__with_invisible(self):
+        class SchemaWithInvisibleColumns(Schema):
+            def __init__(self, schema, visible):
```

Comment:
Let's add an optional `visible` parameter to `MappingSchema` and have it do exactly what you do here.

Then we don't need this test class and the `MappingSchema` class becomes a more full implementation of the interface.